packages/typegpu/src/smol/wgslGenerator.ts

```diff
@@ -159,9 +215,45 @@ function generateExpression(
 
   if ('n' in expression) {
     // Numeric Literal
+    const value = expression.n;
 
-    // TODO: Infer numeric data type from literal
-    return { value: expression.n, dataType: UnknownData };
+    // Hex literals (since JS does not have float hex literals, we'll assume it's an int)
+    const hexRegex = /^0x[0-9a-f]+$/i;
+    if (hexRegex.test(value)) {
+      return { value: `${value}`, dataType: d.abstractInt };
+    }
+
+    // Binary literals
+    const binRegex = /^0b[01]+$/i;
+    if (binRegex.test(value)) {
+      // Since wgsl doesn't support binary literals, we'll convert it to a decimal number
+      return {
+        value: `${Number.parseInt(value.slice(2), 2)}`,
+        dataType: d.abstractInt,
+      };
+    }
+
+    const floatRegex = /^-?(?:\d+\.\d*|\d*\.\d+)$/;
+    if (floatRegex.test(value)) {
+      return { value: `${value}`, dataType: d.abstractFloat };
```

Comment:
```suggestion
      return { value, dataType: d.abstractFloat };
```

same everywhere below